adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,251 @@
+package unicorn
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/buger/jsonparser"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+// unicornImpExt is imp ext for UNICORN
+type unicornImpExt struct {
+	Context *unicornImpExtContext     `json:"context,omitempty"`
+	Bidder  openrtb_ext.ExtImpUnicorn `json:"bidder"`
+}
+
+type unicornImpExtContext struct {
+	Data interface{} `json:"data,omitempty"`
+}
+
+// unicornExt is ext for UNICORN
+type unicornExt struct {
+	Prebid    *openrtb_ext.ExtImpPrebid `json:"prebid,omitempty"`
+	AccountID int64                     `json:"accountId,omitempty"`
+}
+
+// Builder builds a new instance of the UNICORN adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// MakeRequests makes the HTTP requests which should be made to fetch bids.
+func (a *adapter) MakeRequests(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var extRegs openrtb_ext.ExtRegs
+	if request.Regs != nil {
+		if request.Regs.COPPA == 1 {
+			return nil, []error{&errortypes.BadInput{
+				Message: "COPPA is not supported",
+			}}
+		}
+		if err := json.Unmarshal(request.Regs.Ext, &extRegs); err == nil {
+			if extRegs.GDPR != nil && (*extRegs.GDPR == 1) {
+				return nil, []error{&errortypes.BadInput{
+					Message: "GDPR is not supported",
+				}}
+			}
+			if extRegs.USPrivacy != "" {
+				return nil, []error{&errortypes.BadInput{
+					Message: "CCPA is not supported",
+				}}
+			}
+		}
+	}
+
+	err := modifyImps(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	var modifiableSource *openrtb.Source
+	if request.Source != nil {
+		modifiableSource = request.Source
+	} else {
+		modifiableSource = &openrtb.Source{}
+	}
+	modifiableSource.Ext = setSourceExt()
```

Comment:
This code does not actually make a copy of the source. Since the variable is defined as a pointer on line 72, it's still pointing to shared memory. You want to something more like this:

```
var modifiableSource openrtb.Source
if request.Source != nil {
	modifiableSource = *request.Source
}
modifiableSource.Ext = setSourceExt()
request.Source = &modifiableSource
```

I mocked out this situation in the Go Playground if you'd like to explore further: https://play.golang.org/p/rVCn1RzHs4g

Reply:
Oh, Sorry.I not understood it and mistaked...

Fixed it.

https://github.com/prebid/prebid-server/pull/1719/commits/fb58a7ee99d689075713ebb414a0f6cfd7b1412c

Reply:
No worries. :)